:rotating_light: Fix gradient checkpointing for several models and improve test robustness

githubnemo opened the pull request:
Support for gradient checkpointing was lost in the major refactoring in PR #38635 and this is the attempt to re-add it.

I extended the tests to
- test `use_reentrant=True` and `False`
- make sure `model.train` is called so that gradient checkpointing works; this is a limiation of the tests currently used by GPTBigCode
- make sure that one (the first) gradient checkpointing layer is called
- make sure that the same non-zero grads are there for normal and checkpointing runs - this is something we tripped over before in PEFT due to the possibly incompletely stored runtime environment in the checkpointed forward step, see also peft#2826

Note that the invocation of `GPTBigCodeBlock.forward` has changed:

- `layer_past` is now passed as a keyword argument so that `GradientCheckpointingLayer.__call__` can see and filter this parameter (`use_reentrant=False` fails otherwise)
- `{encoder_}hidden_states` are still passed as positional arguments so that `torch.utils.checkpoint.checkpoint` receives them as pos. args and computes gradients for these (kwargs would be filtered by `GradientCheckpointingLayer`).

:rotating_light: Note that this is breaking compatibility by changing the forward signature in `GPTBigCodeBlock.forward`!

Commit:
Implement gradient checkpointing in GPTBigCode

Support for gradient checkpointing was lost in the major refactoring in PR #38635
and this is the attempt to re-add it.

I extended the tests to
- test `use_reentrant=True` and `False`
- make sure `model.train` is called so that gradient checkpointing works;
  this is a limiation of the tests currently used by GPTBigCode
- make sure that one (the first) gradient checkpointing layer is called
- make sure that the same non-zero grads are there for normal and checkpointing
  runs - this is something we tripped over before in PEFT due to the possibly
  incompletely stored runtime environment in the checkpointed forward step,
  see also peft#2826

Note that the invocation of `GPTBigCodeBlock.forward` has changed:

- `layer_past` is now passed as a keyword argument so that
  `GradientCheckpointingLayer.__call__` can see and filter this parameter
  (`use_reentrant=False` fails otherwise)
- `{encoder_}hidden_states` are still passed as positional arguments
  so that `torch.utils.checkpoint.checkpoint` receives them as pos. args
  and computes gradients for these (kwargs would be filtered by
  `GradientCheckpointingLayer`).

vasqu commented on tests/models/gpt_bigcode/test_modeling_gpt_bigcode.py:
```diff
@@ -307,18 +308,43 @@ def create_and_check_lm_head_model(self, config, input_ids, input_mask, token_ty
         self.parent.assertEqual(result.logits.shape, (self.batch_size, self.seq_length, self.vocab_size))
 
     def create_and_check_forward_and_backwards(
-        self, config, input_ids, input_mask, token_type_ids, *args, gradient_checkpointing=False
```
I'm a bit surprised that it was overriden here. It would be nicer if we could move this into `test_modeling_common` instead.

vasqu replied:
We can check if the layer exists somewhere, no? If we do not detect that it exists, raise an error and check which models fails --> all models should already have this but bigger PRs apparently clashed so it isnt the case anymore

vasqu reviewed (commented):
The tests are neat, I think we should move them to common tests tho. Not exactly sure why it was specially treated here. 

And ig there will be a need for another round to check similar models that may have been accidentally overriden with the ckpting layer :sweat: not necessarily this PR tho

vasqu commented on src/transformers/models/gpt_bigcode/modeling_gpt_bigcode.py:
```diff
@@ -291,9 +292,9 @@ def __init__(self, config, layer_idx=None):
+        encoder_hidden_states: Optional[torch.Tensor] = None,
         layer_past: Optional[Cache] = None,
         attention_mask: Optional[torch.Tensor] = None,
-        encoder_hidden_states: Optional[torch.Tensor] = None,
```
I mean that the signature should stay the same, e.g. see https://github.com/huggingface/transformers/blob/84d19be41e0131e6f2a660fe6af8b77094906af7/src/transformers/models/gpt_bigcode/modeling_gpt_bigcode.py#L586-L596

It will need to adjust the calls from the module above like https://github.com/huggingface/transformers/blob/84d19be41e0131e6f2a660fe6af8b77094906af7/src/transformers/models/gpt_bigcode/modeling_gpt_bigcode.py#L901-L910

Changing the signature is breaking a bit too much!

vasqu replied:
For viz, as discussed internally, we need this to be breaking

Commit:
Improve gradient checkpointing tests

- Compare that the non-zero gradients in a reference run are present in the checkpointing run
- Make sure that the forward of at least one gradient checkpointing layer is actually called
  more than once (as expected during gradient checkpointing backward)

Currently there are some problems with Bert-derived MultipleChoice models, when dropout is
enabled there are scenarios during gradient checkpointing where `classifier.bias.grad` is None.
I don't yet have a good explanation for this, disabling dropout resolves this. I would have
understood, if it is dropout on the classification layer but enabling attention dropout is
also leading to this behavior.

MoE models have selective sparsity depending on the selected experts, for this reason we
only compare gradients on parameters collected on the reference backward run.

vasqu reviewed (commented):
I think this is fine, we will likely need to break other model's signature as well? I.e. not only got bigcode. This PR will get bigger than initially thought but let's fix these models

We can allow this for v5 but let's also mention this PR in the v5 thread (#40822) when we merge this.

Commit:
Remove duplicated gradient checkpointing code

Commit:
Address review comments

Commit:
Make test output consistent

githubnemo commented on tests/test_modeling_common.py:
```diff
@@ -1151,22 +1159,64 @@ def check_training_gradient_checkpointing(self, gradient_checkpointing_kwargs=No
+                if hasattr(config, "hidden_dropout_prob"):
+                    config.hidden_dropout_prob = 0.0
+                if hasattr(config, "attention_probs_dropout_prob"):
+                    config.attention_probs_dropout_prob = 0.0
```
> This is only for the multiple choice class? Or are other model types also affected?

It didn't seem to make a difference for other models with the limited test runs I made. If you want I can limit it to the problematic model class.

Commit:
GradientCheckpointingLayer for xlstm, zamba, zamba2

Commit:
GradientCheckpointingLayer for swiftformer

also drop janus from ignore list - only the VQVAE case is without
gradient checkpointing and it is doubtful that it is usefule in that
case. Training with gradient checkpointing is not tested anyway.

vasqu reviewed (commented):
Just noticed this small thing in xlstm

Re: Clvp, let's isolate it for now. We can come back later except you have a good idea how to refactor/handle this properly

githubnemo commented on src/transformers/models/xlstm/modeling_xlstm.py:
```diff
@@ -22,17 +22,21 @@
+
+    class xLSTMBlock(GradientCheckpointingLayer, mLSTMBlock):
+        pass
+
```
No, `mLSTMBlock` doesn't implement `GradientCheckpointingLayer`: https://github.com/NX-AI/xlstm/blob/main/xlstm/xlstm_large/model.py#L455

Commit:
Make an exception for CLVP

The implementation of GradientCheckpointingLayers is not trivial and may break behavior
that was previously expected. Therefore we keep it as-is for now.

vasqu commented on tests/test_modeling_common.py:
```diff
@@ -827,6 +829,17 @@ def test_gradient_checkpointing_enable_disable(self):
+                "janus_vqgan",
+                "clvp",
+                "clvp_encoder",
+                "clvp_decoder",
```
Would it be possible to override this test in `test_modeling_clvp` instead? No biggie if not

githubnemo replied:
I've removed some exceptions (janus, clvp in training test) but for clvp/clvp_decoder I think it is better to have the one single exception visible instead of duplicating the test code without the single assertion. It also didn't make sense to me to refactor the assertion into an abstract base method for the tests since it is only CLVP and nothing else.

vasqu replied:
Gotcha, thanks for iterating here and bearing with me.

vasqu reviewed (approved):
Let's also change the title here a bit since the scope changed

cc @ArthurZucker @Cyrilvallez if you can take a last look --> fixes a last few gradient ckpting models and makes the test more robust towards actually using it properly

Commit:
Remove unneeded exceptions

Cyrilvallez committed:
Merge branch 'main' into issue/gpt-bigcode-gradient-checkpointing

vasqu reviewed (approved):
LGTM, cc @Cyrilvallez for core maintainer

vasqu commented:
Can we also change the title of the PR tho. The PR also makes the test more robust and properly checks for gradient ckpting capabilities

Cyrilvallez reviewed (approved):
Not much to complain about, thanks a lot for this!
